In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Data/laptops.csv')
df

brand             model  ram hd_type  hd_size  screen_size   price  \
0     Dell  Inspiron 15-3567    4     hdd     1024         15.6   40000   
1    Apple       MacBook Air    8     ssd      128         13.3   55499   
2    Apple       MacBook Air    8     ssd      256         13.3   71500   
3    Apple       MacBook Pro    8     ssd      128         13.3   96890   
4    Apple       MacBook Pro    8     ssd      256         13.3  112666   
..     ...               ...  ...     ...      ...          ...     ...   
162   Asus            A555LF    8     hdd     1024         15.6   39961   
163   Asus     X555LA-XX172D    4     hdd      500         15.6   28489   
164   Asus            X554LD    2     hdd      500         15.6   29199   
165   Asus    X550LAV-XX771D    2     hdd      500         15.6   29990   
166   Asus     X540LA-XX538T    4     hdd     1024         15.6   30899   

    processor_brand processor_model  clock_speed graphic_card_brand  \
0             intel              i5          2.5              intel   
1             intel              i5          1.8              intel   
2             intel              i5          1.8              intel   
3             intel              i5          2.3              intel   
4             intel              i5          2.3              intel   
..              ...             ...          ...                ...   
162           intel      i3 4th gen          1.7             nvidia   
163           intel      i3 4th gen          1.9              intel   
164           intel      i3 4th gen          1.9              intel   
165           intel      i3 4th gen          1.7              intel   
166           intel      i3 5th gen          2.0              intel   

     graphic_card_size       os  weight comments  
0                  NaN    linux    2.50      NaN  
1                  2.0      mac    1.35      NaN  
2                  2.0      mac    1.35      NaN  
3                  2.0      mac    3.02      NaN  
4                  2.0      mac    3.02      NaN  
..                 ...      ...     ...      ...  
162                2.0  windows    2.30      NaN  
163                NaN    linux    2.30      NaN  
164                1.0    linux    2.30      NaN  
165                NaN    linux    2.50      NaN  
166                NaN  windows    2.30      NaN  

[167 rows x 15 columns]

In [3]:
df.tail(6)

brand           model  ram hd_type  hd_size  screen_size  price  \
161  Asus  X541UA-GO1345D    4     hdd     1024         15.6  27730   
162  Asus          A555LF    8     hdd     1024         15.6  39961   
163  Asus   X555LA-XX172D    4     hdd      500         15.6  28489   
164  Asus          X554LD    2     hdd      500         15.6  29199   
165  Asus  X550LAV-XX771D    2     hdd      500         15.6  29990   
166  Asus   X540LA-XX538T    4     hdd     1024         15.6  30899   

    processor_brand processor_model  clock_speed graphic_card_brand  \
161           intel      i3 6th gen          2.0              intel   
162           intel      i3 4th gen          1.7             nvidia   
163           intel      i3 4th gen          1.9              intel   
164           intel      i3 4th gen          1.9              intel   
165           intel      i3 4th gen          1.7              intel   
166           intel      i3 5th gen          2.0              intel   

     graphic_card_size       os  weight comments  
161                NaN    linux    1.84      NaN  
162                2.0  windows    2.30      NaN  
163                NaN    linux    2.30      NaN  
164                1.0    linux    2.30      NaN  
165                NaN    linux    2.50      NaN  
166                NaN  windows    2.30      NaN

In [4]:
df.head(6)

brand                   model  ram hd_type  hd_size  screen_size   price  \
0   Dell        Inspiron 15-3567    4     hdd     1024         15.6   40000   
1  Apple             MacBook Air    8     ssd      128         13.3   55499   
2  Apple             MacBook Air    8     ssd      256         13.3   71500   
3  Apple             MacBook Pro    8     ssd      128         13.3   96890   
4  Apple             MacBook Pro    8     ssd      256         13.3  112666   
5  Apple  MacBook Pro (TouchBar)   16     ssd      512         15.0  226000   

  processor_brand processor_model  clock_speed graphic_card_brand  \
0           intel              i5          2.5              intel   
1           intel              i5          1.8              intel   
2           intel              i5          1.8              intel   
3           intel              i5          2.3              intel   
4           intel              i5          2.3              intel   
5           intel              i7          2.7              intel   

   graphic_card_size     os  weight comments  
0                NaN  linux    2.50      NaN  
1                2.0    mac    1.35      NaN  
2                2.0    mac    1.35      NaN  
3                2.0    mac    3.02      NaN  
4                2.0    mac    3.02      NaN  
5                2.0    mac    2.50      NaN

In [5]:
df.shape

(167, 15)

데이터의 크기를 나타냄

In [6]:
df.columns

Index(['brand', 'model', 'ram', 'hd_type', 'hd_size', 'screen_size', 'price',
       'processor_brand', 'processor_model', 'clock_speed',
       'graphic_card_brand', 'graphic_card_size', 'os', 'weight', 'comments'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   brand               167 non-null    object 
 1   model               167 non-null    object 
 2   ram                 167 non-null    int64  
 3   hd_type             167 non-null    object 
 4   hd_size             167 non-null    int64  
 5   screen_size         167 non-null    float64
 6   price               167 non-null    int64  
 7   processor_brand     167 non-null    object 
 8   processor_model     167 non-null    object 
 9   clock_speed         166 non-null    float64
 10  graphic_card_brand  163 non-null    object 
 11  graphic_card_size   81 non-null     float64
 12  os                  167 non-null    object 
 13  weight              160 non-null    float64
 14  comments            55 non-null     object 
dtypes: float64(4), int64(3), object(8)
memory usage: 19.7+ KB

각 컬럼의 개수 167 -> 오른쪽은 데이터 타입 나타냄

로우 수는 167이지만 값이 비어있을 수도 있으므로 167보다 작은 수를 갖는 컬럼이 있기도 한 것임

In [8]:
df.describe()

ram     hd_size  screen_size          price  clock_speed  \
count  167.000000   167.00000   167.000000     167.000000   166.000000   
mean     6.898204   768.91018    14.775210   64132.898204     2.321084   
std      3.787479   392.99080     1.376526   42797.674010     0.554187   
min      2.000000    32.00000    10.100000   13872.000000     1.100000   
25%      4.000000   500.00000    14.000000   35457.500000     1.900000   
50%      8.000000  1024.00000    15.600000   47990.000000     2.300000   
75%      8.000000  1024.00000    15.600000   77494.500000     2.600000   
max     16.000000  2048.00000    17.600000  226000.000000     3.800000   

       graphic_card_size      weight  
count          81.000000  160.000000  
mean           52.160494    2.250812  
std           444.134142    0.648446  
min             1.000000    0.780000  
25%             2.000000    1.900000  
50%             2.000000    2.200000  
75%             4.000000    2.600000  
max          4000.000000    4.200000

각 컬럼별 통계값을 나타냄

In [9]:
df.sort_values(by = 'price')

brand                       model  ram hd_type  hd_size  screen_size  \
148       Acer             Aspire SW3-016     2     ssd       32         10.1   
83        Acer     A315-31CDC UN.GNTSI.001    2     ssd      500         15.6   
108       Acer  Aspire ES-15 NX.GKYSI.010     4     hdd      500         15.6   
100       Acer    A315-31-P4CRUN.GNTSI.002    4     hdd      500         15.6   
73        Acer             Aspire ES1-523     4     hdd     1024         15.6   
..         ...                         ...  ...     ...      ...          ...   
154  Microsoft      Surface Book CR9-00013    8     ssd      128         13.5   
31        Acer                 Predator 17   16     ssd      256         17.3   
96   Alienware          AW13R3-7000SLV-PUS    8     ssd      256         13.3   
90   Alienware                 15 Notebook   16     hdd     1024         15.6   
5        Apple      MacBook Pro (TouchBar)   16     ssd      512         15.0   

      price processor_brand processor_model  clock_speed graphic_card_brand  \
148   13872           intel      Atom Z8300         1.44              intel   
83    17990           intel         Celeron         1.10              intel   
108   17990             amd         A4-7210         1.80                amd   
100   18990           intel         pentium         1.10              intel   
73    19465             amd         A4-7210         1.80                amd   
..      ...             ...             ...          ...                ...   
154  178799           intel              i5         1.80              intel   
31   178912           intel              i7         2.60             nvidia   
96   190256           intel              i7         3.00             nvidia   
90   199000           intel              i7         2.60             nvidia   
5    226000           intel              i7         2.70              intel   

     graphic_card_size       os  weight  \
148                NaN  windows     1.2   
83                 NaN  windows     2.1   
108                NaN  windows     2.4   
100                NaN  windows     NaN   
73                 NaN    linux     2.4   
..                 ...      ...     ...   
154                NaN  windows     1.5   
31                 NaN  windows     4.2   
96                 6.0  windows     2.6   
90                 8.0  windows     3.5   
5                  2.0      mac     2.5   

                                              comments  
148                                                NaN  
83                                                 NaN  
108                                                NaN  
100                                                NaN  
73                                                 NaN  
..                                                 ...  
154                                                NaN  
31                                 Integrated Graphics  
96   13.3 inch FHD (1920 x 1080) IPS Anti-Glare 300...  
90      Maximum Display Resolution : 1920 x 1080 pixel  
5                                                  NaN  

[167 rows x 15 columns]

sort_values 하고 by price를 하면 가격을 기준으로 가격이 작은 것부터 나열됨

In [10]:
df.sort_values(by = 'price', ascending = False)

brand                       model  ram hd_type  hd_size  screen_size  \
5        Apple      MacBook Pro (TouchBar)   16     ssd      512         15.0   
90   Alienware                 15 Notebook   16     hdd     1024         15.6   
96   Alienware          AW13R3-7000SLV-PUS    8     ssd      256         13.3   
31        Acer                 Predator 17   16     ssd      256         17.3   
154  Microsoft      Surface Book CR9-00013    8     ssd      128         13.5   
..         ...                         ...  ...     ...      ...          ...   
73        Acer             Aspire ES1-523     4     hdd     1024         15.6   
100       Acer    A315-31-P4CRUN.GNTSI.002    4     hdd      500         15.6   
108       Acer  Aspire ES-15 NX.GKYSI.010     4     hdd      500         15.6   
83        Acer     A315-31CDC UN.GNTSI.001    2     ssd      500         15.6   
148       Acer             Aspire SW3-016     2     ssd       32         10.1   

      price processor_brand processor_model  clock_speed graphic_card_brand  \
5    226000           intel              i7         2.70              intel   
90   199000           intel              i7         2.60             nvidia   
96   190256           intel              i7         3.00             nvidia   
31   178912           intel              i7         2.60             nvidia   
154  178799           intel              i5         1.80              intel   
..      ...             ...             ...          ...                ...   
73    19465             amd         A4-7210         1.80                amd   
100   18990           intel         pentium         1.10              intel   
108   17990             amd         A4-7210         1.80                amd   
83    17990           intel         Celeron         1.10              intel   
148   13872           intel      Atom Z8300         1.44              intel   

     graphic_card_size       os  weight  \
5                  2.0      mac     2.5   
90                 8.0  windows     3.5   
96                 6.0  windows     2.6   
31                 NaN  windows     4.2   
154                NaN  windows     1.5   
..                 ...      ...     ...   
73                 NaN    linux     2.4   
100                NaN  windows     NaN   
108                NaN  windows     2.4   
83                 NaN  windows     2.1   
148                NaN  windows     1.2   

                                              comments  
5                                                  NaN  
90      Maximum Display Resolution : 1920 x 1080 pixel  
96   13.3 inch FHD (1920 x 1080) IPS Anti-Glare 300...  
31                                 Integrated Graphics  
154                                                NaN  
..                                                 ...  
73                                                 NaN  
100                                                NaN  
108                                                NaN  
83                                                 NaN  
148                                                NaN  

[167 rows x 15 columns]

ascending = False를 하면 값이 큰 것부터 아래로 나열됨

In [11]:
df.sort_values(by = 'price', ascending = False, inplace = True)

In [12]:
df

brand                       model  ram hd_type  hd_size  screen_size  \
5        Apple      MacBook Pro (TouchBar)   16     ssd      512         15.0   
90   Alienware                 15 Notebook   16     hdd     1024         15.6   
96   Alienware          AW13R3-7000SLV-PUS    8     ssd      256         13.3   
31        Acer                 Predator 17   16     ssd      256         17.3   
154  Microsoft      Surface Book CR9-00013    8     ssd      128         13.5   
..         ...                         ...  ...     ...      ...          ...   
73        Acer             Aspire ES1-523     4     hdd     1024         15.6   
100       Acer    A315-31-P4CRUN.GNTSI.002    4     hdd      500         15.6   
108       Acer  Aspire ES-15 NX.GKYSI.010     4     hdd      500         15.6   
83        Acer     A315-31CDC UN.GNTSI.001    2     ssd      500         15.6   
148       Acer             Aspire SW3-016     2     ssd       32         10.1   

      price processor_brand processor_model  clock_speed graphic_card_brand  \
5    226000           intel              i7         2.70              intel   
90   199000           intel              i7         2.60             nvidia   
96   190256           intel              i7         3.00             nvidia   
31   178912           intel              i7         2.60             nvidia   
154  178799           intel              i5         1.80              intel   
..      ...             ...             ...          ...                ...   
73    19465             amd         A4-7210         1.80                amd   
100   18990           intel         pentium         1.10              intel   
108   17990             amd         A4-7210         1.80                amd   
83    17990           intel         Celeron         1.10              intel   
148   13872           intel      Atom Z8300         1.44              intel   

     graphic_card_size       os  weight  \
5                  2.0      mac     2.5   
90                 8.0  windows     3.5   
96                 6.0  windows     2.6   
31                 NaN  windows     4.2   
154                NaN  windows     1.5   
..                 ...      ...     ...   
73                 NaN    linux     2.4   
100                NaN  windows     NaN   
108                NaN  windows     2.4   
83                 NaN  windows     2.1   
148                NaN  windows     1.2   

                                              comments  
5                                                  NaN  
90      Maximum Display Resolution : 1920 x 1080 pixel  
96   13.3 inch FHD (1920 x 1080) IPS Anti-Glare 300...  
31                                 Integrated Graphics  
154                                                NaN  
..                                                 ...  
73                                                 NaN  
100                                                NaN  
108                                                NaN  
83                                                 NaN  
148                                                NaN  

[167 rows x 15 columns]

기존의 데이터 자체를 정렬하고 싶으면 inplace = True를 사용하면 됨

In [13]:
df['brand']

5          Apple
90     Alienware
96     Alienware
31          Acer
154    Microsoft
         ...    
73          Acer
100         Acer
108         Acer
83          Acer
148         Acer
Name: brand, Length: 167, dtype: object

In [14]:
df['brand'].unique()

array(['Apple', 'Alienware', 'Acer', 'Microsoft', 'HP', 'Dell', 'Lenovo',
       'Asus'], dtype=object)

unique 메소드를 이용하면 겹치는 것을 제외한 브랜드를 알 수 있음

In [15]:
df['brand'].value_counts()

HP           55
Acer         35
Dell         31
Lenovo       18
Asus          9
Apple         7
Microsoft     6
Alienware     6
Name: brand, dtype: int64

value_counts는 각 브랜드가 몇 번씩 나오는 지를 알 수 있음

In [16]:
df['brand'].describe()

count     167
unique      8
top        HP
freq       55
Name: brand, dtype: object

브랜드의 길이는 167, 서로 다른 브랜드 개수는 8개, 가장 자주 등장하는 브랜드(top)는 HP, 그 빈도수는 55

In [17]:
world_df = pd.read_csv('Data/world_cities.csv')
world_df

Unnamed: 0 City / Urban area     Country  Population  Land area (in sqKm)
0             0      Buenos Aires   Argentina    11200000                 2266
1             1         Melbourne   Australia     3162000                 2080
2             2            Sydney   Australia     3502000                 1687
3             3          Brisbane   Australia     1508000                 1603
4             4             Perth   Australia     1177000                  964
..          ...               ...         ...         ...                  ...
244         245            Canton         USA      267000                  372
245         246           Spokane         USA      335000                  371
246         247          Tashkent  Uzbekistan     2200000                  531
247         248  Ho Chi Minh City     Vietnam     4900000                  518
248         249            Harare    Zimbabwe     1750000                  712

[249 rows x 5 columns]

In [18]:
world_df['Country'].value_counts()

USA            105
France          15
Brazil          10
Canada           9
Germany          8
              ... 
Vietnam          1
Azerbaijan       1
Ghana            1
Indonesia        1
Philippines      1
Name: Country, Length: 61, dtype: int64

In [19]:
world_df['Country'].value_counts().shape

(61,)

In [20]:
world_df.columns

Index(['Unnamed: 0', 'City / Urban area', 'Country', 'Population',
       'Land area (in sqKm)'],
      dtype='object')

In [21]:
world_df['density'] = world_df['Population'] / world_df['Land area (in sqKm)']

In [22]:
df_high_density = world_df[world_df['density']> 10000] 

In [23]:
df_high_density.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19 entries, 32 to 129
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           19 non-null     int64  
 1   City / Urban area    19 non-null     object 
 2   Country              19 non-null     object 
 3   Population           19 non-null     int64  
 4   Land area (in sqKm)  19 non-null     int64  
 5   density              19 non-null     float64
dtypes: float64(1), int64(3), object(2)
memory usage: 1.0+ KB


In [24]:
ranks = world_df.sort_values(by = 'density', ascending = False)
ranks

Unnamed: 0 City / Urban area   Country  Population  Land area (in sqKm)  \
75           75            Mumbai     India    14350000                  484   
74           74           Kolkata     India    12700000                  531   
101         101           Karachi  Pakistan     9800000                  518   
99           99             Lagos   Nigeria    13400000                  738   
34           34          Shenzhen     China     8000000                  466   
..          ...               ...       ...         ...                  ...   
194         195       Chattanooga       USA      344000                  751   
223         224         Asheville       USA      222000                  536   
57           57               Pau    France      181000                  450   
220         221           Hickory       USA      188000                  546   
196         197   Barnstable Town       USA      244000                  741   

          density  
75   29648.760331  
74   23917.137476  
101  18918.918919  
99   18157.181572  
34   17167.381974  
..            ...  
194    458.055925  
223    414.179104  
57     402.222222  
220    344.322344  
196    329.284750  

[249 rows x 6 columns]

In [25]:
ranks['City / Urban area']

75              Mumbai
74             Kolkata
101            Karachi
99               Lagos
34            Shenzhen
            ...       
194        Chattanooga
223          Asheville
57                 Pau
220            Hickory
196    Barnstable Town
Name: City / Urban area, Length: 249, dtype: object

In [26]:
world_df['Country']

0       Argentina
1       Australia
2       Australia
3       Australia
4       Australia
          ...    
244           USA
245           USA
246    Uzbekistan
247       Vietnam
248      Zimbabwe
Name: Country, Length: 249, dtype: object

In [27]:
countries = world_df['Country'].value_counts()
countries > 4

USA             True
France          True
Brazil          True
Canada          True
Germany         True
               ...  
Vietnam        False
Azerbaijan     False
Ghana          False
Indonesia      False
Philippines    False
Name: Country, Length: 61, dtype: bool

In [28]:
countries[countries == 4]

Italy    4
Name: Country, dtype: int64

country가 네 번 나왔다는 것은 도시가 4개라는 뜻이므로 4번 나온 국가를 찾아주면 된다.

countries[countries == 4] 이렇게 쓰면 4일 때 불린 값이 True인 로우(countries == 4가 True인 경우)만 나오게 됨

In [29]:
countries[countries == 4].index

Index(['Italy'], dtype='object')

In [30]:
countries = world_df['Country'].value_counts() > 5

In [37]:
df.loc[countries, 'Population'] = 0

IndexingError: (USA             True
France          True
Brazil          True
Canada          True
Germany         True
               ...  
Vietnam        False
Azerbaijan     False
Ghana          False
Indonesia      False
Philippines    False
Name: Country, Length: 61, dtype: bool, 'Population')